components/clarity-lsp/src/common/requests/definitions.rs

```diff
@@ -0,0 +1,289 @@
+    ) -> bool {
+        self.visit_define_function_parameters(parameters)
+    }
+
+    fn visit_match_option(
+        &mut self,
+        expr: &'a SymbolicExpression,
+        _input: &'a SymbolicExpression,
+        some_name: &'a ClarityName,
+        _some_branch: &'a SymbolicExpression,
+        _none_branch: &'a SymbolicExpression,
+    ) -> bool {
+        let declaration = expr
+            .match_list()
+            .expect("expr should be list")
+            .get(2)
+            .expect("match should have declaration");
+        self.definitions
+            .insert(some_name.to_owned(), span_to_range(&declaration.span));
+        true
+    }
+
+    fn visit_match_response(
+        &mut self,
+        expr: &'a SymbolicExpression,
+        _input: &'a SymbolicExpression,
+        ok_name: &'a ClarityName,
+        _ok_branch: &'a SymbolicExpression,
+        err_name: &'a ClarityName,
+        _err_branch: &'a SymbolicExpression,
+    ) -> bool {
+        let ok_declaration = expr
+            .match_list()
+            .expect("expr should be list")
+            .get(2)
+            .expect("match should have declaration");
+        self.definitions
+            .insert(ok_name.to_owned(), span_to_range(&ok_declaration.span));
+
+        let err_declaration = expr
+            .match_list()
+            .expect("expr should be list")
+            .get(4)
+            .expect("match should have declaration");
+        self.definitions
+            .insert(err_name.to_owned(), span_to_range(&err_declaration.span));
+        true
+    }
+
+    fn visit_let(
+        &mut self,
+        _expr: &'a SymbolicExpression,
+        bindings: &HashMap<&'a ClarityName, &'a SymbolicExpression>,
+        _body: &'a [SymbolicExpression],
+    ) -> bool {
+        for (name, expr) in bindings {
+            self.definitions
+                .insert((*name).to_owned(), span_to_range(&expr.span));
+        }
+        true
+    }
+}
+
+fn find_local_definition(
+    search: &ClarityName,
+    contract_location: &FileLocation,
+    expressions: &Vec<SymbolicExpression>,
+    position: &Position,
+) -> Option<Location> {
+    let let_visitor = LocalBindingsVisitor::new(position);
+    let let_definitions = let_visitor.get_local_bindings(&expressions);
+    if let Some(range) = let_definitions.get(search) {
+        return Some(Location::new(
+            Url::parse(&contract_location.to_string()).ok()?,
+            *range,
+        ));
+    }
+    None
+}
+
+pub fn find_definition(
+    location: &FileLocation,
+    position: &Position,
+    expressions: &Vec<SymbolicExpression>,
+) -> Option<Location> {
+    let search = get_expression_name_at_position(position, expressions, None)?;
+
+    if let Some(location) = find_global_definition(&search, location, expressions) {
+        return Some(location);
+    }
+
+    if let Some(location) = find_local_definition(&search, location, expressions, position) {
+        return Some(location);
+    }
```

Comment:
Instead of traversing the AST each time, could you traverse once and save a set of scopes and their definitions, then quickly reference that for each query?